libraries/persistent_store/fuzz/src/store.rs

```diff
@@ -0,0 +1,533 @@
+        cycle: usize,
+    },
+}
+
+impl Init {
+    /// Returns whether fuzzing is in dirty mode.
+    fn is_dirty(&self) -> bool {
+        match self {
+            Init::Dirty => true,
+            _ => false,
+        }
+    }
+
+    /// Returns the number of used cycles.
+    ///
+    /// This is zero if the storage was not artificially aged.
+    fn used_cycles(&self) -> usize {
+        match self {
+            Init::Used { cycle } => *cycle,
+            _ => 0,
+        }
+    }
+}
+
+/// Compact stack of bits.
+// NOTE: This would probably go away once the delay map is simplified.
+#[derive(Default, Clone, Debug)]
+struct BitStack {
+    /// Bits stored in little-endian (for bytes and bits).
+    data: Vec<u8>,
+
+    /// Number of bits stored.
+    len: usize,
+}
+
+impl BitStack {
+    /// Returns whether the stack is empty.
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns the length of the stack.
+    fn len(&self) -> usize {
+        if self.len == 0 {
+            8 * self.data.len()
+        } else {
+            8 * (self.data.len() - 1) + self.len
+        }
+    }
+
+    /// Pushes a bit to the stack.
+    fn push(&mut self, value: bool) {
+        if self.len == 0 {
+            self.data.push(0);
+        }
+        if value {
+            *self.data.last_mut().unwrap() |= 1 << self.len;
+        }
+        self.len += 1;
+        if self.len == 8 {
+            self.len = 0;
+        }
+    }
+
+    /// Pops a bit from the stack.
+    fn pop(&mut self) -> Option<bool> {
+        if self.len == 0 {
+            if self.data.is_empty() {
+                return None;
+            }
+            self.len = 8;
+        }
+        self.len -= 1;
+        let result = self.data.last().unwrap() & 1 << self.len;
+        if self.len == 0 {
+            self.data.pop().unwrap();
+        }
+        Some(result != 0)
+    }
+}
+
+impl std::fmt::Display for BitStack {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        let mut bits = self.clone();
+        while let Some(bit) = bits.pop() {
+            write!(f, "{}", bit as usize)?;
+        }
+        write!(f, " ({} bits)", self.len())?;
+        Ok(())
+    }
+}
+
+#[test]
+fn bit_stack_ok() {
+    let mut bits = BitStack::default();
+
+    assert_eq!(bits.pop(), None);
+
+    bits.push(true);
+    assert_eq!(bits.pop(), Some(true));
+    assert_eq!(bits.pop(), None);
+
+    bits.push(false);
+    assert_eq!(bits.pop(), Some(false));
+    assert_eq!(bits.pop(), None);
+
+    bits.push(true);
+    bits.push(false);
+    assert_eq!(bits.pop(), Some(false));
+    assert_eq!(bits.pop(), Some(true));
+    assert_eq!(bits.pop(), None);
+
+    bits.push(false);
+    bits.push(true);
+    assert_eq!(bits.pop(), Some(true));
+    assert_eq!(bits.pop(), Some(false));
+    assert_eq!(bits.pop(), None);
+
+    for i in 0..27 {
+        assert_eq!(bits.len(), i);
+        bits.push(true);
+    }
```

Comment:
Maybe `pop` them all back? Since you never `pop` over a byte border.

Reply:
Good point. Done.